tools/hsmtools.c

```diff
@@ -282,6 +286,119 @@ static int dump_commitments_infos(struct node_id *node_id, u64 channel_id,
+ * hash specified in the witness programm.
+ *
+ * :param address: The bech32 address of the P2WPKH output
+ * :param node_id: The id of the node with which the channel was established
+ * :param depth: How many commitments to derive for each dbid (the basepoint
+ *               is checked by default).
+ * :param hsm_secret_path: The path to the hsm_secret
+ * :param passwd: The *optional* hsm_secret password
+ */
+static int guess_commitment(const char *address, struct node_id *node_id,
+                            u64 depth, char *hsm_secret_path, char *passwd)
+{
+	struct sha256 shaseed;
+	struct secret hsm_secret, channel_seed, basepoint_secret;
+	struct privkey payment_privkey;
+	struct pubkey per_commitment_point, basepoint, payment_pubkey;
+	struct ripemd160 pubkeyhash;
+	/* We only support P2WPKH, hence 20 */
+	u8 goal_pubkeyhash[20];
+	/* See common/bech32.h for buffer size */
+	char hrp[strlen(address) - 6];
+	int witver;
+	size_t witlen;
+
+	/* Get the hrp to accept addresses from any network */
+	if (bech32_decode(hrp, goal_pubkeyhash, &witlen, address, 90) != 1)
+		errx(ERROR_USAGE, "Could not get address' network");
+	if (segwit_addr_decode(&witver, goal_pubkeyhash, &witlen, hrp, address) != 1)
+		errx(ERROR_USAGE, "Wrong bech32 address");
+
+	secp256k1_ctx = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY
+	                                         | SECP256K1_CONTEXT_SIGN);
+
+	if (passwd)
+		get_encrypted_hsm_secret(&hsm_secret, hsm_secret_path, passwd);
+	else
+		get_hsm_secret(&hsm_secret, hsm_secret_path);
+
+	for (u64 dbid = 0; ; dbid++) {
+		get_channel_seed(&channel_seed, node_id, dbid, &hsm_secret);
+		if (!derive_payment_basepoint(&channel_seed,
+		                              &basepoint, &basepoint_secret))
+			errx(ERROR_KEYDERIV, "Could not derive basepoints for dbid %"PRIu64
+			                     " and channel seed %s.", dbid,
+			                     type_to_string(tmpctx,
+			                                    struct secret, &channel_seed));
+
+		/* If we specified option_static_remotekey, this is just the basepoint ! */
+		pubkey_to_hash160(&basepoint, &pubkeyhash);
+		if (memcmp(pubkeyhash.u.u8, goal_pubkeyhash, 20) == 0) {
+			printf("FOUND !\n");
+			printf("pubkey hash : %s\n",
+			       tal_hexstr(tmpctx, pubkeyhash.u.u8, 20));
+			printf("pubkey      : %s \n",
+			       type_to_string(tmpctx, struct pubkey, &basepoint));
+			printf("privkey     : %s \n",
+			       type_to_string(tmpctx, struct secret, &basepoint_secret));
+			return 0;
+		}
+
+		/* Otherwise derive payment keys up to <depth> to check */
+		if (!derive_shaseed(&channel_seed, &shaseed))
+			errx(ERROR_KEYDERIV, "Could not derive shaseed for dbid %"PRIu64
+			                     " and channel seed %s.", dbid,
+			                     type_to_string(tmpctx,
+			                                    struct secret, &channel_seed));
+		for (u64 i = 0; i < depth; i++) {
```

Comment:
I'm wrong on this branch since I can't manage to recover the payment key without `option_static_remotekey`.... Can't manage to find what's wrong though..

Reply:
Ok after hours of testing I realized that we actually derive the `to_remote`'s pubkey from OUR commitment point, which means we can't recover the `to_remote` outputs to us without knowing the remote peer's `per_commit_point`..................